sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/OfferEvaluator.java

```diff
@@ -353,92 +341,100 @@ static void logOutcome(StringBuilder stringBuilder, EvaluationOutcome outcome, S
-            String taskName = entry.getKey();
-            List<ResourceSpec> resourceSpecs = getOrderedResourceSpecs(entry.getValue());
-
-            for (ResourceSpec resourceSpec : resourceSpecs) {
-                if (resourceSpec instanceof NamedVIPSpec) {
-                    evaluationStages.add(new NamedVIPEvaluationStage(
-                            (NamedVIPSpec) resourceSpec, taskName, Optional.empty(), resourceNamespace));
-                } else if (resourceSpec instanceof PortSpec) {
-                    evaluationStages.add(new PortEvaluationStage(
-                            (PortSpec) resourceSpec, taskName, Optional.empty(), resourceNamespace));
-                } else {
-                    evaluationStages.add(new ResourceEvaluationStage(
-                            resourceSpec, Optional.of(taskName), Optional.empty(), resourceNamespace));
-                }
-
-                if (preReservedRole == null && role == null && principal == null) {
-                    preReservedRole = resourceSpec.getPreReservedRole();
-                    role = resourceSpec.getRole();
-                    principal = resourceSpec.getPrincipal();
-                }
-            }
-
-            for (VolumeSpec volumeSpec : entry.getValue().getVolumes()) {
-                evaluationStages.add(VolumeEvaluationStage.getNew(
-                        volumeSpec, Optional.of(taskName), resourceNamespace));
+        Map<String, ResourceSet> resourceSetsByTaskSpecName =
+                podInstanceRequirement.getPodInstance().getPod().getTasks().stream()
+                // Create a TreeMap: Doesn't hurt to have consistent ordering when evaluating tasks
+                .collect(Collectors.toMap(
+                        TaskSpec::getName,
+                        TaskSpec::getResourceSet,
+                        (u, v) -> {
+                            throw new IllegalStateException(String.format("Duplicate key %s", u));
+                        },
+                        TreeMap::new));
+
+        // Only reserve the executor's resources once:
+        boolean addedExecutorResources = false;
+        // For any given ResourceSet, only reserve its configured resources once (but store TaskInfos for all tasks):
+        Set<String> addedResourceSets = new HashSet<>();
+
+        for (Map.Entry<String, ResourceSet> taskEntry : resourceSetsByTaskSpecName.entrySet()) {
+            String taskSpecName = taskEntry.getKey();
+            ResourceSet resourceSet = taskEntry.getValue();
+            List<ResourceSpec> resourceSpecs = getOrderedResourceSpecs(resourceSet);
+
+            if (!addedExecutorResources) {
+                // The default executor needs a fixed amount of "overhead" resources, to be added once per pod.
+                // For consistency, let's put this before the per-task/resourceset RESERVE calls below.
+                addedExecutorResources = true;
+
+                getExecutorResourceSpecs(
```

Comment:
Since `getExecutorResourceSpecs` is private, we could change that to rather return the list of `ResourceEvaluationStage`s instead of performing the map here.